### ISSUE TITLE
Node 16 / Electron 15 prebuilds

### ISSUE THREAD

**Issue by dopry** [MadLittleMods/node-usb-detection#153]
It looks like prebuilds are being run with node 14. Node 16 is current and Electron 15 has shipped with Node 16.  It would be nice to have prebuilts for the current version of node/electron.

**Comment by mholtzman** [comment]
> Looks like the root issue is [electron/electron-rebuild#886](https://github.com/electron/electron-rebuild/issues/886). The fix has been merged, but a new version has not yet been cut.

I was looking at this earlier and, while this should work for allowing electron-rebuild to recognize the correct ABI version, it ultimately still calls prebuild-install for projects like this so I don't think this will solve it.

**Comment by mcous** [comment]
@mholtzman yeah, I think you're right. The resolution override only works because it's overriding it in two places. That's... not ideal.

@dopry of course! I think this ticket should stay open / possibly pinned, since until upstream issues are resolved, users will be unable to install this package in Electron v15 without the above workarounds

**Comment by Stringer86** [comment]
Followed the above instructions (and package.json) but still getting:

/node_modules/usb-detection/build/Release/detection.node'
was compiled against a different Node.js version using
NODE_MODULE_VERSION 89. This version of Node.js requires
NODE_MODULE_VERSION 98. Please try re-compiling or re-installing
the module (for instance, using `npm rebuild` or `npm install`).
    at process.func [as dlopen] (node:electron/js2c/asar_bundle:5:1800)
    at Object.Module._extensions..node (node:internal/modules/cjs/loader:1170:18)
    at Object.func [as .node] (node:electron/js2c/asar_bundle:5:1800)
    at Module.load (node:internal/modules/cjs/loader:982:32)
    at Module._load (node:internal/modules/cjs/loader:823:12)
    at Function.c._load (node:electron/js2c/asar_bundle:5:13331)
    at Module.require (node:internal/modules/cjs/loader:1006:19)
    at require (node:internal/modules/cjs/helpers:93:18)

"usb-detection": "^4.13.0"
"electron": "^15.2.0",
"electron-rebuild": "^3.2.3",
"electron-builder": "^22.14.5",

Any advice?

**Comment by mholtzman** [comment]
@dopry it looks like prebuild-install was updated to include a more current version of node-abi, so I think this issue could be fixed by updating to use [prebuild-install 7.x](https://github.com/prebuild/prebuild-install/releases/tag/v7.0.0)

**Comment by mholtzman** [comment]
> Followed the above instructions (and package.json) but still getting:
> 
> ```
> /node_modules/usb-detection/build/Release/detection.node'
> was compiled against a different Node.js version using
> NODE_MODULE_VERSION 89. This version of Node.js requires
> NODE_MODULE_VERSION 98. Please try re-compiling or re-installing
> the module (for instance, using `npm rebuild` or `npm install`).
>     at process.func [as dlopen] (node:electron/js2c/asar_bundle:5:1800)
>     at Object.Module._extensions..node (node:internal/modules/cjs/loader:1170:18)
>     at Object.func [as .node] (node:electron/js2c/asar_bundle:5:1800)
>     at Module.load (node:internal/modules/cjs/loader:982:32)
>     at Module._load (node:internal/modules/cjs/loader:823:12)
>     at Function.c._load (node:electron/js2c/asar_bundle:5:13331)
>     at Module.require (node:internal/modules/cjs/loader:1006:19)
>     at require (node:internal/modules/cjs/helpers:93:18)
> ```
> 
> ```
> "usb-detection": "^4.13.0"
> "electron": "^15.2.0",
> "electron-rebuild": "^3.2.3",
> "electron-builder": "^22.14.5",
> ```
> 
> Any advice?

you used npm-force-resolutions and it still doesn't work? check your package-lock.json file to make sure everything is using node-abi 3.x+

**Comment by sfoster1** [comment]
I'm not sure whether this is part of your configurations or an upstream thing in `node-gyp` or `nan` but v8 requires C++17 to build it properly since the v8 9.3 release ([this](https://chromium.googlesource.com/v8/v8/+/8fd20298e5e9fdb69614ba37be1cba8930049dd9) is the commit that did it). V8 9.3 is used in [Node 16.9](https://nodejs.org/ja/blog/release/v16.9.0/) and later. Most compilers unfortunately don't yet default to C++17, so when your code includes `node.h` you eventually get errors like this during the build:

```
npm ERR! /home/seth/.electron-gyp/20.3.0/include/node/v8-maybe.h:106:45: error: ‘is_lvalue_reference_v’ is not a member of ‘std’; did you mean ‘is_lvalue_reference’?
    npm ERR!   106 |   template <class U, std::enable_if_t<!std::is_lvalue_reference_v<U>>*>
    npm ERR!       |                                             ^~~~~~~~~~~~~~~~~~~~~
    npm ERR!       |                                             is_lvalue_reference
    npm ERR! /home/seth/.electron-gyp/20.3.0/include/node/v8-maybe.h:106:66: warning: logical not is only applied to the left hand side of comparison [-Wlogical-not-parentheses]
    npm ERR!   106 |   template <class U, std::enable_if_t<!std::is_lvalue_reference_v<U>>*>
    npm ERR!       |                                                                  ^
    npm ERR! /home/seth/.electron-gyp/20.3.0/include/node/v8-maybe.h:106:67: error: template argument 1 is invalid
    npm ERR!   106 |   template <class U, std::enable_if_t<!std::is_lvalue_reference_v<U>>*>
    npm ERR!       |                                                                   ^
    npm ERR! /home/seth/.electron-gyp/20.3.0/include/node/v8-maybe.h:106:71: error: expected unqualified-id before ‘>’ token
    npm ERR!   106 |   template <class U, std::enable_if_t<!std::is_lvalue_reference_v<U>>*>
    npm ERR!       |                                                                       ^
    npm ERR! /home/seth/.electron-gyp/20.3.0/include/node/v8-maybe.h:123:43: error: ‘is_lvalue_reference_v’ is not a member of ‘std’; did you mean ‘is_lvalue_reference’?
    npm ERR!   123 | template <class T, std::enable_if_t<!std::is_lvalue_reference_v<T>>* = nullptr>
    npm ERR!       |                                           ^~~~~~~~~~~~~~~~~~~~~
    npm ERR!       |                                           is_lvalue_reference
    npm ERR! /home/seth/.electron-gyp/20.3.0/include/node/v8-maybe.h:123:64: warning: logical not is only applied to the left hand side of comparison [-Wlogical-not-parentheses]
    npm ERR!   123 | template <class T, std::enable_if_t<!std::is_lvalue_reference_v<T>>* = nullptr>
    npm ERR!       |                                                                ^
    npm ERR! /home/seth/.electron-gyp/20.3.0/include/node/v8-maybe.h:123:65: error: template argument 1 is invalid
    npm ERR!   123 | template <class T, std::enable_if_t<!std::is_lvalue_reference_v<T>>* = nullptr>
    npm ERR!       |                                                                 ^
    npm ERR! /home/seth/.electron-gyp/20.3.0/include/node/v8-maybe.h:123:70: error: expected unqualified-id before ‘=’ token
    npm ERR!   123 | template <class T, std::enable_if_t<!std::is_lvalue_reference_v<T>>* = nullptr>
```

(I know this is in the context of electron, but it's a change in v8 that would happen in any node distribution >=16.9).

If you run `CPPFLAGS="-std=c++17" (npm|yarn) install`, or if you downgrade to node 16.8, the build succeeds fine.

**Comment by sfoster1** [comment]
Also fyi: https://github.com/electron/electron/issues/35193 affects this module too - anything that uses `nan` and targets recent node
